### PR TITLE
WIP: Add initial support for Blazor WebAssembly

### LINKED ISSUE #13755: Microsoft.PowerShell.SDK doesn't work in a AspNetCore Blazor WebAssembly project
**transferred from** [dotnet/aspnetcore#25844](https://github.com/dotnet/aspnetcore/issues/25844)

When using the preview of AspNetCore Blazor WebAssembly and attempting to utilize the Microsoft.PowerShell.SDK package, Blazor WebAssembly is unable to load the System.Management.Automation.dll

# Steps to reproduce
* Create new Blazor WASM project. (net5-preview.7)
* Update project file according to [this guide.](https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-5-preview-7/)
* Add `Microsoft.PowerShell.SDK` package reference
* Do something with a `PowerShell` host [Example repo](https://github.com/jrhodnik/PowerShellTests)

## Expected behavior

it works

## Actual behavior

Exceptions: 
```none
Could not load file or assembly 'System.Management.Automation, Version=7.0.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies.
```

## Environment data

* PowerShell 7.0.3
* AspNetCore Blazor WebAssembly 5.0 Preview 7

### PR DESCRIPTION
**Pull request by fMichaleczek**
# PR Summary

- Declaring OperatingSystem.IsBrowser in the startup process
- Add a minimal logger support through Console.WriteLine
- Add a sample demo of a Blazor WebAssembly with SMA 

## PR Context

Fixes #13755

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: demos/Blazor/powershell-blazor-wasm-demo/wwwroot/index.html [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<!DOCTYPE html>
+<html>
+
+<head>
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" />
+    <title>PowerShell Demo</title>
+    <base href="/" />
+    <link href="css/bootstrap/bootstrap.min.css" rel="stylesheet" />
```
Comment (TravisEz13) [inline demos/Blazor/powershell-blazor-wasm-demo/wwwroot/index.html:9]: We should have a mention of the license and copyright for these files.

File: demos/Blazor/powershell-blazor-wasm-demo/wwwroot/css/open-iconic/FONT-LICENSE [hunk 2]
```diff
@@ -0,0 +1,86 @@
+SIL OPEN FONT LICENSE Version 1.1
```
Comment (TravisEz13) [inline demos/Blazor/powershell-blazor-wasm-demo/wwwroot/css/open-iconic/FONT-LICENSE:1]: Compliance review needed

File: demos/Blazor/powershell-blazor-wasm-demo/wwwroot/css/open-iconic/ICON-LICENSE [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
```
Comment (TravisEz13) [inline demos/Blazor/powershell-blazor-wasm-demo/wwwroot/css/open-iconic/ICON-LICENSE:21]: add newline to make posix compliant

**Review by TravisEz13 — CHANGES_REQUESTED** [review]
Personally, I think this should be in a separate repo.
I'd rather not confuse people with license for the repo vs the product.

**Comment by TravisEz13** [comment]
cc @SteveL-MSFT Please comment if demos that add licenses should be in the same or different repos.

**Comment by fMichaleczek** [comment]
@TravisEz13 

> The demo folder is more for reviewers if they want to see live working sample,  it's the default blazor webassembly template from the Asp.Net Core SDK. I am OK to remove it from the final review. I will publish a real sample later and we will discuss about it.
> I will remove the PowerShell Logger part in the PR because I don't need anymore, the boot process is clean and without error (Blazor already provide a logger for the rest). I will wait for an general logger improvement in 7.2 timeline.

**Comment by SteveL-MSFT** [comment]
It may make sense to have a powershell-blazor repo for the demo files which may evolve into something more useful.  @fMichaleczek if you're ok with that, I can create the repo and you can submit a PR for the demo files there and just have this PR be PowerShell changes needed to run in browser

**Comment by fMichaleczek** [comment]
@SteveL-MSFT This PR is a POC to say : it works, see by yourself. First, PowerShell need to be fileless in Blazor environment. I can first make a PR for the version git number and some path. Then it will be about your plugin system, do we go to DI like all dotnet projet for conf, log etc ? Microsoft.Extensions ? Finally we have to update the buildchain (to build on windows), the tests and the demos (maybe better outside). Then we will have an infrastructure and I will begin the RFC to ask community. Do you want more details about a particular item ?

**Comment by SteveL-MSFT** [comment]
@fMichaleczek "fileless" meaning it has to be self contained and not try to access the file system?  We have a RFC coming for pluggable subsystems for PowerShell and maybe logging should be a subsystem and can be replaced.  What I'm specifically looking for is changes to PowerShell so that it can be hosted in blazor.  Just like how we have a console host today, I'd imagine there would be a separate blazor host which would be its own project/repo.

**Comment by fMichaleczek** [comment]
@SteveL-MSFT 

> "fileless" meaning it has to be self contained and not try to access the file system?  

When you load SMA, the inmemory filesystem is empty. json, psd1, psm1 files are unavailable at starting. You can only use assemblies. or download the files and have a lot of latencies. 

> We have a RFC coming for pluggable subsystems for PowerShell and maybe logging should be a subsystem and can be replaced.  

args (maybe not apply to all cases), configuration file, logging and the way we load modules have to evolve.

> What I'm specifically looking for is changes to PowerShell so that it can be hosted in blazor.  

We are looking for changes to PowerShell so that it can be hosted in AspNetCore, Blazor Wasm, Blazor Wasm Server and MAUI Blazor Wasm.

- PowerShell is always loaded after the DI, so I need a fileless way to configure and inject SMA with a smallset of modules.

- webclient is broken and http is the only  way  to communicate

- MMI package is still weird for unknow platform

> Just like how we have a console host today, I'd imagine there would be a separate blazor host which would be its own project/repo.

I have a monacoeditor + xterm.js demo to finish. 200Mo, need a 100Mbits network 🙃 SMA + SDK are to big because of dependencies.

Maybe better to continue here :
https://github.com/PowerShell/PowerShell/issues/13755

**Pull request closed without merging**